### DISCUSSION TITLE
Assigning class weights to imbalanced dataset

### DISCUSSION THREAD

**Discussion by keerti-ds** [ultralytics/yolov5#5604] [Q&A] [answered]
Hi, I am new to Pytroch and I have a difficulty in understanding the concept of setting class weights for imbalanced dataset.

I know I can set class weights in Tensorflow and Keras using `from sklearn.utils import class_weight` as 

`model.fit(X_train, Y_train, nb_epoch=5, batch_size=32, class_weight=class_weight)`

I would like to know if this `labels_to_class_weights` function is for the same purpose? Because I see that they are normalizing the weights. I am quite confused, is it not possible to assign the weights as we do in Tensorlfow like?

`class_weight = class_weight.compute_class_weight('balanced'
                                               ,np.unique(Y_train)
                                               ,Y_train)`

or this function is doing exactly the same thing?

https://github.com/ultralytics/yolov5/blob/7ebb5e5da673350d4c168cf60d01986a5e0f00cc/utils/general.py#L495

Thanks in advance.

**Comment by glenn-jocher** [comment]
@keerti-ds class weights are computed automatically for all datasets in train.py here:
https://github.com/ultralytics/yolov5/blob/7ebb5e5da673350d4c168cf60d01986a5e0f00cc/train.py#L255

They're not used during training by default, but you can do whatever you want with them.

**Comment by keerti-ds** [comment, marked as answer]
@glenn-jocher Thanks for the reply. What do I need to change to use them for training? Because I don't see any parameter where I can pass them.

**Reply by glenn-jocher** [reply]
CELoss and BCELoss functions have weight arguments. You might experiment with using these weights in classification loss in loss.py
https://pytorch.org/docs/stable/generated/torch.nn.BCELoss.html

https://github.com/ultralytics/yolov5/blob/7ebb5e5da673350d4c168cf60d01986a5e0f00cc/utils/loss.py#L91-L101

**Comment by keerti-ds** [comment]
I see BCEWithLogitsLoss also has a weight argument. Can I not use this? Is there any particular reason you suggested me BCELoss or CELoss?

And if I would want to pass the class weights, can I just add them in `hyp.yaml` and access in `loss.py` from this file as you are accessing `h['cls_pw']`?

**Reply by glenn-jocher** [reply]
@keerti-ds yes sure you can do whatever you want.

pw is a positive weight, which is different than a class weight. See the PyTorch docs.

**Reply by kalfasyan** [reply]
The original question is how to use class_weights and I don't see how it got answered in this discussion. How do we use the model's class weights during training?

**Reply by glenn-jocher** [reply]
@kalfasyan CELoss and BCELoss functions have weight arguments. You might experiment with using these weights in classification loss in loss.py
https://pytorch.org/docs/stable/generated/torch.nn.BCELoss.html

https://github.com/ultralytics/yolov5/blob/7ebb5e5da673350d4c168cf60d01986a5e0f00cc/utils/loss.py#L91-L101

**Comment by keerti-ds** [comment]
Thanks! But I see weight argument in docs. 

`torch.nn.BCEWithLogitsLoss(weight=None, size_average=None, reduce=None, reduction='mean', pos_weight=None)`
[https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html#torch.nn.BCEWithLogitsLoss](url)

And pw is here `pos_weight` if I understand correctly.  And `weight` should be the class weight. Please correct me if I am wrong.

And thanks, I will try with other losses.

**Reply by glenn-jocher** [reply]
@keerti-ds as far as I understand it pos_weight is a scalar and weight may be a vector. The two arguments serve two completely different purposes. But I'm not the module author, if you have questions regarding it you may want to raise them on the pytorch forums.

**Comment by RicoSuaveGuapo** [comment]
Correct me if I was wrong

Authors provide sampling method for imbalance dataset through `--image-weights` flag in train.py. One can check https://github.com/ultralytics/yolov5/blob/63ddb6f0d06f6309aa42bababd08c859197a27af/train.py#L288-L291
where they use class weights and instances' classes in a image to decide how dataloader sample data (since this satistics needs dataloader to be static, so they use Dataloader not InfiniteDataloader). In the other word, sampling more images with rare classes than the others.

The other method is directly change the class weight which is this thread mainly discussed. We can get the class weights directly from authors' code 
https://github.com/ultralytics/yolov5/blob/63ddb6f0d06f6309aa42bababd08c859197a27af/train.py#L266
with the shape of (nc). One can save/copy it, then put it to `hyp.scratch.yaml`'s option `cls_pw`.
To be clearify, https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html
says that
* `pos_weight`: has to shape of classes, which should a vector
* `weight`: is to control how you weight elements in a batch. I.e, for a batch [tensor_a, tensor_b], we set weight=[10, 1] means you handle the loss computation with additional 10 times weight for tensor_a

**Comment by saulcanortiz** [comment]
Hi!
Then, imagine you are training an imbalanced custom dataset with 4 classes (3 minor classes and 1 major class) and you want to set this class_weights = [0.3, 0.3, 0.3, 0.1]. How can you do it using any yolov5 sub-model? @glenn-jocher

**Reply by glenn-jocher** [reply]
`model.class_weights` are automatically created from your dataset based on inverse frequency. This can be used to customize loss functions etc, but are not used by default.

**Reply by saulcanortiz** [reply]
Therefore, if I declare model.class_weights = torch.tensor([0.3, 0.3, 0.3, 0.1]), the alpha weights of the Focal loss in losses.py are going to be the model.class_weights. Is it correct? Should I add extra lines of code to make the alpha-weights = model.cass_weights?

Regards

**Reply by glenn-jocher** [reply]
@saulcanortiz yes you can access your model.class_weights inside the loss functions in utils/loss.py and apply them however you want.
